### ISSUE TITLE
Does HPC workflow - always require MPI?

### ISSUE THREAD

**Issue by dougbenjamin** [PanDAWMS/pilot2#140]
In the wiki documentation - 

The Pilot 2 HPC workflow is a special mode where the application works without a remote connection to PanDA server or other remote facilities. All intercommunications in this case are managed by the Harvester application. Also, in this mode Pilot 2 acts like a simple MPI application, which performs execution of multiple jobs on the computing nodes of the HPC.

What about HPC jobs that do not require MPI?

**Comment by DanilaOleynik** [comment]
It's a quite abstract question, I think.  
Jobs, themselves is should not be an MPI application (at least we had no chance to test it with typical ATLAS workloads). HPC Workflow is an MPI application, which executes one PanDA job per one MPI rank, without expecting intercommunication between ranks. If the number of ranks will be 1 - it will be one job per pilot. That does not mean, that workloads will be restricted for use MPI messages if needed. Also, into account should be taken the mode of Harvester:    "ManyToOne" mode is expected.
May I ask for an example of:  "HPC jobs that do not require MPI"?

**Comment by dougbenjamin** [comment]
Danila,

   It is not an abstract question at all.

  In fact you have answered the question"HPC Workflow is an MPI application,
which executes one PanDA job per one MPI rank,  without expecting
intercommunication between ranks."

As I have seen in the code, rank information is not used.

If there is no communication between ranks then is absolutely no need for
the MPI libraries, it is an unneeded dependency.

The only ATLAS HPC workflow that requires MPI is Yoda/Droid event service
running.

Examples of HPC jobs that do not require MPI,

1) all of the minipilot running at NERSC last year
2) your running at OLCF
3) ANALY GPU running at the BNL IC cluster.

Doug

**Comment by DanilaOleynik** [comment]
Doug, 

The rank number is used for association with PanDA Job and allows to track a particular node with a PanDA job.

https://github.com/PanDAWMS/pilot2/blob/649b061cb7ca057b9fb04b74583ac81df156c901/pilot/resource/titan.py#L29 

In simple words - there is a list of JobIDs associated with 'worker', and rank 0 will be the first job from this list, rank 2 - second etc.   

By some reason, I had to put it into resources specific plugin. I will need to remember why I did it in this way...
